deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -257,62 +251,219 @@ def _copy_all_manifest_files_from_old_hash_buckets(
+
+    Yields:
+        Tuple[List, List]: A tuple containing two lists:
+            1. A list of DeltaFileEnvelope objects representing UPSERT operations.
+            2. A list of DeltaFileEnvelope objects representing DELETE operations.
+    """
+    iter_df_envelopes = iter(df_envelopes)
+    upserts, deletes = [], []
+    for is_upsert, group_envelopes in itertools.groupby(
+        iter_df_envelopes, lambda x: x.delta_type == DeltaType.UPSERT
+    ):
+        if is_upsert:
+            upserts.extend(group_envelopes)
+        else:
+            deletes.extend(group_envelopes)
+        if deletes:
+            yield upserts, deletes
+            upserts, deletes = [], []
+    if upserts:
+        yield upserts, []
+
+
+def _compact_table_v2(
+    input: MergeInput,
+    dfe_list: List[List[DeltaFileEnvelope]],
+    hb_idx: int,
+    has_deletes: bool,
+) -> Tuple[pa.Table, int, int, int]:
+    """
+    Compacts a list of DeltaFileEnvelope objects into a single PyArrow table.
+
+    Args:
+        input (MergeInput): The input for the merge operation.
+        dfe_list (List[List[DeltaFileEnvelope]]): A list of lists of DeltaFileEnvelope objects.
+        hb_idx (int): The hash bucket index for the compaction.
+        has_deletes (bool): Whether the input data contains DELETE operations.
+
+    Returns:
+        Tuple[pa.Table, int, int, int]: A tuple containing:
+            1. The compacted PyArrow table.
+            2. The total number of records in the incremental data.
+            3. The total number of deduplicated records.
+            4. The total number of dropped records due to DELETE operations.
+    """
+    df_envelopes: List[DeltaFileEnvelope] = _flatten_dfe_list(dfe_list)
+    delete_file_envelopes = input.delete_file_envelopes or []
+    all_dfes: List[DeltaFileEnvelope] = _sort_df_envelopes(
+        delete_file_envelopes + df_envelopes
+    )
+    assert all(
+        dfe.delta_type in (DeltaType.UPSERT, DeltaType.DELETE) for dfe in all_dfes
+    ), "All reordered delta file envelopes must of the UPSERT or DELETE"
+    prev_table, table = None, None
+    aggregated_incremental_len = 0
+    aggregated_deduped_records = 0
+    aggregated_dropped_records = 0
+    if _has_previous_compacted_table(input, hb_idx):
+        table = _download_compacted_table(
+            hb_index=hb_idx,
+            rcf=input.round_completion_info,
+            read_kwargs_provider=input.read_kwargs_provider,
+            deltacat_storage=input.deltacat_storage,
+            deltacat_storage_kwargs=input.deltacat_storage_kwargs,
+        )
+        if has_deletes:
+            table, partial_dropped_rows = input.delete_strategy.apply_all_deletes(
+                table, input.delete_file_envelopes
+            )
+            aggregated_dropped_records += partial_dropped_rows
+    prev_table = table
+    for i, (upsert_group, delete_group) in enumerate(
+        _group_by_upsert_delete_sequence(all_dfes)
+    ):
```

Comment:
Should we process one group at once? When processing upsert we call apply_upersts and during deletes we call apply_deletes?

Reply:
I don't follow.  `_group_by_upsert_delete_sequence(all_dfes)` is an iterator that yields a sequence of Upserts bookended by Deletes if present. Once we have the list of upsert and deletes we are operating on list of upserts and deletes DeltaFileEnvelopes (not individual elements). How would you refactor it? 

```
for i, (upsert_group, delete_group) in enumerate(
        _group_by_upsert_delete_sequence(all_dfes)
    ):
        current_upsert_dfe_sequence: List[DeltaFileEnvelope] = list(upsert_group)
        current_delete_dfe_sequence: List[DeltaFileEnvelope] = list(delete_group)
         if current_upsert_dfe_sequence:
               _apply_upserts()
         if current_delete_dfe_sequence:
              input.delete_strategy.apply_many_deletes()
```

Reply:
Do you have to group upserts with deletes? You could iterate on group of upserts and deletes separately.

Reply:
Changed
```
def _compact_table_v2(
    input: MergeInput,
    dfe_list: List[List[DeltaFileEnvelope]],
    hb_idx: int,
    has_deletes: bool,
    compacted_table: Optional[pa.Table] = None,
) -> Tuple[pa.Table, int, int, int]:
    """
    Compacts a list of DeltaFileEnvelope objects into a single PyArrow table.

    Args:
        input (MergeInput): The input for the merge operation.
        dfe_list (List[List[DeltaFileEnvelope]]): A list of lists of DeltaFileEnvelope objects.
        hb_idx (int): The hash bucket index for the compaction.
        has_deletes (bool): Whether the input data contains DELETE operations.

    Returns:
        Tuple[pa.Table, int, int, int]: A tuple containing:
            1. The compacted PyArrow table.
            2. The total number of records in the incremental data.
            3. The total number of deduplicated records.
            4. The total number of dropped records due to DELETE operations.
    """
    df_envelopes: List[DeltaFileEnvelope] = _flatten_dfe_list(dfe_list)
    delete_file_envelopes = input.delete_file_envelopes or []
    reordered_all_dfes: List[DeltaFileEnvelope] = _sort_df_envelopes(
        delete_file_envelopes + df_envelopes
    )
    assert all(
        dfe.delta_type in (DeltaType.UPSERT, DeltaType.DELETE) for dfe in reordered_all_dfes
    ), "All reordered delta file envelopes must of the UPSERT or DELETE"
    prev_table, table = compacted_table, compacted_table
    aggregated_incremental_len = 0
    aggregated_deduped_records = 0
    aggregated_dropped_records = 0
    for i, (delta_type, delta_type_sequence) in enumerate(_group_by_sequence(reordered_all_dfes)):
        if delta_type is DeltaType.UPSERT:
            (
                table,
                partial_incremental_len,
                partial_deduped_records,
                partial_merge_time,
            ) = _apply_upserts(input, delta_type_sequence, hb_idx, prev_table)
            logger.info(
                f"[Merge task index {input.merge_task_index}] Merged "
                f"record count: {len(table)}, size={table.nbytes} took: {partial_merge_time}s"
            )
            aggregated_incremental_len += partial_incremental_len
            aggregated_deduped_records += partial_deduped_records
        elif delta_type is DeltaType.DELETE:
            (table, dropped_rows) = input.delete_strategy.apply_many_deletes(
                table, delta_type_sequence
            )
            logger.info(
                f"[Merge task index {input.merge_task_index}] Dropped "
                + f"record count: {dropped_rows}"
            )
            aggregated_dropped_records += dropped_rows
        prev_table = table
    return (
        table,
        aggregated_incremental_len,
        aggregated_deduped_records,
        aggregated_dropped_records,
    )
```